### PR TITLE
Fix #44068 (thick smoke causes character sprite to disappear completely) + more

### LINKED ISSUE #44068: Thick smoke causes character sprite to disappear completely
# Describe the bug

When standing in the thick smoke (e.g. near the thick smoke vent), smoke gets so thick, that character sprite disappears.

# Steps To Reproduce

1. Debug smoke vent (3) near you
2. Stand in the smoke (may require to skip several turns)
3. Char disappears

# Expected behavior

Character sprite (and a tile under character) should be always visible.

# Screenshots

smoke

# Versions and configuration

- OS: MacOs
    - OS Version: Name: Mac OS X; Version: 10.14; Build: 18A391; 
- Game Version: 0.E-5816-ge1277a9f38-dirty [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Magiclysm [magiclysm]
]

# Additional context

This behavior is also reproducible in version 446c7504733b51b0178292ae104fae6edf0b0551 , before my recent lighting-related PRs.

### PR DESCRIPTION
**Pull request by Aivean**
#### Summary

SUMMARY:  Bugfixes "Several transparency related issues"

#### Purpose of change

When testing #44093 I've uncovered two existing issues related to transparency and visibility. 
First one is described here: #44068.  Another issue is related, character sprite disappears in following scenario:
`visibility_variables::vision_threshold` returns `LIGHT_AMBIENT_LOW` and [this strict check fails](https://github.com/CleverRaven/Cataclysm-DDA/compare/master...Aivean:fix-thick-smoke-obscures-character?expand=1#diff-b5681127b2f10bf71edd2778a617486cL735).

#### Describe the solution

I did the following changes:

* [fixed transparency override](https://github.com/CleverRaven/Cataclysm-DDA/compare/master...Aivean:fix-thick-smoke-obscures-character?expand=1#diff-b5681127b2f10bf71edd2778a617486cR178) in `vision_transparency_cache` using correct transparency value
* [safeguarded](https://github.com/CleverRaven/Cataclysm-DDA/compare/master...Aivean:fix-thick-smoke-obscures-character?expand=1#diff-b5681127b2f10bf71edd2778a617486cR341) against transparency values below  `LIGHT_TRANSPARENCY_OPEN_AIR` that caused sun light to grow uncontrollable
* [used the correct cache](https://github.com/CleverRaven/Cataclysm-DDA/compare/master...Aivean:fix-thick-smoke-obscures-character?expand=1#diff-b5681127b2f10bf71edd2778a617486cR634) in the `apparent_light_at` calculation (this was causing thick smoke bug, as transparency override was not working)
* [made](https://github.com/CleverRaven/Cataclysm-DDA/compare/master...Aivean:fix-thick-smoke-obscures-character?expand=1#diff-b5681127b2f10bf71edd2778a617486cR733) `a.apparent_light >= cache.vision_threshold`  comparison not strict to allow `LIGHT_AMBIENT_LOW` at zero range be visible
* [removed](https://github.com/CleverRaven/Cataclysm-DDA/compare/master...Aivean:fix-thick-smoke-obscures-character?expand=1#diff-167344106cab6092dca1c0fc9c159e26L25) wrong and confusing transparency constant `LIGHT_TRANSPARENCY_CLEAR`
* added some documentation on transparency and caches 

#### Describe alternatives you've considered

I guess there were other places where it was possible to override lighting and transparency behavior for player's tile. I tried to stay as noninvasive as possible.

#### Testing

I manually playtested multiple scenarios involving various light levels, different z-levels, perception skill level, etc., even created debug transparency overlay #44108 to better understand how transparency works and interacts with light. I'm pretty confident that nothing is broken by this change.

#### Additional context

Before:
bug2

After:
fixed1

### CODE REVIEW

**Comment by Aivean** [comment]
I forgot to remove `LIGHT_TRANSPARENCY_CLEAR` references from the tests, working on it.

File: src/lightmap.cpp [hunk 1]
```diff
@@ -340,9 +338,9 @@ void map::build_sunlight_cache( int pzlev )
                     if( prev_transparency > LIGHT_TRANSPARENCY_SOLID &&
                         !prev_floor_cache[prev_x][prev_y] &&
                         ( prev_light_max = prev_lm[prev_x][prev_y].max() ) > 0.0 ) {
-                        const float light_level = std::max( inside_light_level, prev_light_max *
+                        const float light_level = std::min( prev_light_max, std::max( inside_light_level, prev_light_max *
```
Comment (AMurkin) [inline src/lightmap.cpp:341]: You can use [clamp](https://github.com/CleverRaven/Cataclysm-DDA/blob/16dfef8d18a9c597a73f8c4f05c941b286514650/src/cata_utility.h#L222) here.

File: tests/vision_test.cpp [hunk 2]
```diff
@@ -545,3 +548,32 @@ TEST_CASE( "vision_see_wall_in_moonlight", "[shadowcasting][vision]" )
...
+            "61116",
+            "61116",
+            "66666",
+        },
+        midnight,
+        vision_test_flags::none
+    };
+
+    t.test_all();
+}
```
Comment (ZhilkinSerg) [inline tests/vision_test.cpp:579]:
```suggestion
}

```

**Review by JulianEmirGR — COMMENTED** [review]
Na

**Pull request merged**